src/Raven.Server/Documents/Queries/QueriedDocument.cs

```diff
@@ -0,0 +1,133 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries;
+
+public sealed class QueriedDocument : Document
+{
+    private static readonly QueriedDocument ExplicitNullReference = new();
+    private int _refCount = 1;
+    public int RefCount => _refCount;
+
+    // Cloned document is not stored in LRU, that means disposing this object should only decrease reference at original document. 
+    private QueriedDocument _sourceDocument;
+
+    // This is used by document that were loaded additionaly to our document from query.
+    private QueriedDocument _referencingSingle;
+    private List<QueriedDocument> _referencing;
+
+    public bool CanDispose => _refCount == 1;
+    public bool IsCloned;
+
+    public void LinkReferencedDocument(QueriedDocument additionalDocument)
```

Comment:
`additionalDocument` is a document that is referenced right? Why not to call to `referencedDocument`?

Reply:
Done